public/app/features/search/service/frontend.ts

```diff
@@ -0,0 +1,239 @@
+
+  // Fields
+  id?: Vector<number>;
+  url?: Vector<string>;
+  uid?: Vector<string>;
+  name?: Vector<string>;
+  description?: Vector<string>;
+  dashboardID?: Vector<number>;
+  type?: Vector<string>;
+  tags?: Vector<string>; // JSON strings?
+}
+
+interface CompositeKey {
+  kind: SearchResultKind;
+  index: number;
+}
+
+export function getFrontendGrafanaSearcher(data: RawIndexData): GrafanaSearcher {
+  const searcher = new MiniSearch<InputDoc>({
+    idField: '__id',
+    fields: ['name', 'description', 'tags'], // fields to index for full-text search
+    searchOptions: {
+      boost: {
+        name: 3,
+        description: 1,
+      },
+      // boost dashboard matches first
+      boostDocument: (documentId: any, term: string) => {
+        const kind = documentId.kind;
+        if (kind === 'dashboard') {
+          return 1.4;
+        }
+        if (kind === 'folder') {
+          return 1.2;
+        }
+        return 1;
+      },
+      prefix: true, // (term) => term.length > 3,
+      fuzzy: (term) => (term.length > 4 ? 0.2 : false),
+    },
+    extractField: (doc, name) => {
+      // return a composite key for the id
+      if (name === '__id') {
+        return {
+          kind: doc.kind,
+          index: doc.index,
+        };
+      }
+      const values = (doc as any)[name] as Vector;
+      if (!values) {
+        return undefined;
+      }
+      return values.get(doc.index);
+    },
+  });
+
+  const lookup = new Map<SearchResultKind, InputDoc>();
+  for (const [key, frame] of Object.entries(data)) {
+    const kind = key as SearchResultKind;
+    const input = getInputDoc(kind, frame);
+    lookup.set(kind, input);
+    for (let i = 0; i < frame.length; i++) {
+      input.index = i;
+      searcher.add(input);
+    }
+  }
+
+  // Construct the URL field for each panel
+  if (true) {
+    const dashboard = lookup.get('dashboard');
+    const panel = lookup.get('panel');
+    if (dashboard?.id && panel?.dashboardID && dashboard.url) {
+      const dashIDToIndex = new Map<number, number>();
+      for (let i = 0; i < dashboard.id?.length; i++) {
+        dashIDToIndex.set(dashboard.id.get(i), i);
+      }
+
+      const urls: string[] = new Array(panel.dashboardID.length);
+      for (let i = 0; i < panel.dashboardID.length; i++) {
+        const dashboardID = panel.dashboardID.get(i);
+        const index = dashIDToIndex.get(dashboardID);
+        if (index != null) {
+          urls[i] = dashboard.url.get(index) + '?viewPanel=' + panel.id?.get(i);
+        }
+      }
+      panel.url = new ArrayVector(urls);
+    }
+  }
+
+  return {
+    search: async (query: string, filter?: QueryFilters) => {
+      const found = searcher.search(query);
+
+      // frame fields
+      const url: string[] = [];
+      const kind: string[] = [];
+      const type: string[] = [];
+      const name: string[] = [];
+      const info: any[] = [];
+      const score: number[] = [];
+
+      for (const res of found) {
```

Comment:
We could probably break this out into a separate util function which would clean this up a bit and also be easier to test

Maybe something like
`
generateResults = (found) => {}
`

then we could just `return generateResults(found)`